### ISSUE TITLE
[Proposal] Tuples as constructors!

### ISSUE THREAD

**Issue by VBAndCs** [dotnet/vblang#554]
As we can deconstruct objects to tuples, I suggest to be able to construct objects from tuples, without explicitly writing a widening operator for that.
For example, we can use:
`Dim p as point = (100, 200)`
and vb should lower that to:
`Dim p as point = new point (100, 200)`
This doesn't seem useful, but it can be when passing params. Suppose we have a method that expects an optional param array of points, then we can use:
`Foo((0,0), (0, 100), (100, 100), (100, 0))`
Instead of:
`Foo(new Point(0,0), new Point(0, 100), new Point(100, 100), new Point(100, 0))`
and the same when we initialize arrays and lists:
`Dim L = new List(of Point) From {(0,0), (0, 100), (100, 100), (100, 0)}`
It is shorter than:
```
Dim L = new List(of Point) From {
new Point(0,0), 
new Point(0, 100), 
new Point(100, 100), 
new Point(100, 0)
}
```

And so on.

In short:
If there is no direct conversion from Tuple to the object, VB should look for a constructor that has the same no of arguments as the tuple, with a matching types, or types that can accept the conversion from the typle's elements types.
The intellisense also should offer the constructors signatures to make typing easier.

**Comment by AdamSpeight2008** [comment]
I think the Roslyn team, call them "Tuple Deconstructors".

**Comment by zspitz** [comment]
@AdamSpeight2008 AFAIK [tuple deconstruction](https://docs.microsoft.com/en-us/dotnet/csharp/deconstruct) is something else -- extracting parts of an object into a value tuple. In the C# implementation, this is done using an instance or extension `Deconstruct` method with the appropriate number of parameters:
```
Class Person
    Property FirstName As String
    Property LastName As String
    Sub Deconstruct(ByRef LastName As String, ByRef FirstName As String)
        LastName = Me.LastName
        FirstName = Me.FirstName
    End Sub
End Class

Dim p As New Person With {.LastName = "Smith", .FirstName = "John"}
Dim (LastName, FirstName) = p
```
This proposal suggests finding an appropriate constructor given the types of a value tuple and calling that constructor with the given values:
```
Class Person
    Property FirstName As String
    Property LastName As String
    Sub New(LastName As String, FirstName As String)
        Me.LastName = LastName
        Me.FirstName = FirstName
    End Sub
End Class

Dim p As Person = ("Smith", "John")
```
A more appropriate name might be "construction by tuple".

**Comment by VBAndCs** [comment]
Why do we need the keyword new?
This is read as an assignment of a tuple to an object, but under the hood, the compiler calls the appropriate constructor.
Before tuples, we might need  the new keyword, since the (, ) syntax had no meaning itself. But after tuples, all it is meaningful, so, why fill the page with unnecessary words?
This is VB after all, not a crippled C-family Language.

**Comment by VBAndCs** [comment]
Yes, this will be valid:
```
Dim data = (100, 0) 'As (Integer, Integer)
Dim p As Point = data
```

The compiler can pass the tuple elements to the appropriate constructor.

**Comment by VBAndCs** [comment]
We should always treat a tuple dynamically as a sequence of data, so, we can pass this sequence between objects and receive it back from them, without any further effort of the programmers.
This is how we used to have VB: a language that do most of the work for us. C# take that away in last years and in turn, VB fade away! I want VB back, with a team with a VB brain.

**Comment by gilfusion** [comment]
Okay. A couple more questions come to mind:

First, it looks like you're basically adding an implicit conversion from a tuple to any type with an appropriate constructor. Would that also work for an explicit conversion (`CType`)?
```vb
Dim info = (100, 0)
```
- Would `CType((100, 0), Point)` be equivalent to `New Point(100, 0)`?
- Would `CType(info, Point)` be equivalent to `New Point(info.Item1, info.Item2)`?

Second, would your proposal also work for zero- and single-argument constructors? (I don't believe zero- or one-element tuples are exactly supported in VB.)
```vb
Dim randomGeneratorPool = New List(Of Random) From { (), (), (), (), () }
```

**Comment by VBAndCs** [comment]
I don't want to focus on CType now, as I don't get the reason of doing that.
In fact I wrote the issue thinking that VB supports deconstrucors, but luckily it doesn't! So, I vote against #346 (I think it is mine with an old account :) ), and propose to use the construvtors also as destrucors in the inverse direction to the one I proposed in this issue!
So, just by having constructors, we can convert from (construct) and to (destruct) Tuples!
This is more superior to C# current ways, as it allows us to use these two features with .NET Core and third party classes that don't have CType operators and Deconstruct methods!
VB has always been a smart language that understands us, and carry out most of the work for us. I want to get this spirit back. This always has a slight  performance penalty, which we accept to pay in RAD, small and test aps. This also makes the language easier for beginners. But we always have the choice to write a longer version of the code with better performance if we need it.
We must get rid of the C# user-unfriendly performance-oriented design. Mangling the two teams corrupted VB and killed it eventually!

**Comment by CyrusNajmabadi** [comment]
Closing this out as this shoudl already be possible today.  You can just provide an implciit widening conversion from the tuple type to your type.  And, if you don't want to bother writing that manually, you could just use SourceGenerators to spit these out for you automatically.
